src/workerd/server/isolate-limit-enforcer.c++

```diff
@@ -0,0 +1,236 @@
+  }
+  void completedRequest(kj::StringPtr id) const override {}
+  bool exitJs(jsg::Lock& lock) const override { return false; }
+  void reportMetrics(IsolateObserver& isolateMetrics) const override {}
+  kj::Maybe<size_t> checkPbkdfIterations(jsg::Lock& lock, size_t iterations) const override {
+    // No limit on the number of iterations in workerd
+    return kj::none;
+  }
+};
+
+class LimitedArrayBufferAllocator final: public v8::ArrayBuffer::Allocator {
+public:
+  LimitedArrayBufferAllocator(size_t limit): limit(limit) {}
+  ~LimitedArrayBufferAllocator() {}
+
+  void* Allocate(size_t length) override {
+    if (length > limit) return nullptr;
+    return calloc(length, 1);
+  }
+  void* AllocateUninitialized(size_t length) override {
+    if (length > limit) return nullptr;
+    return malloc(length);
+  }
+  void Free(void* data, size_t length) override {
+    free(data);
+  }
+
+private:
+  size_t limit;
+};
+
+class ConfiguredIsolateLimitEnforcer final: public IsolateLimitEnforcer {
+public:
+  ConfiguredIsolateLimitEnforcer(server::config::Worker::Limits::Reader limits)
+      : softHeapLimitMb(limits.getHeapSoftLimitMb()),
+        heapHardLimitMb(limits.getHeapHardLimitMb()),
+        heapSnapshotNearHeapLimit(limits.getHeapSnapshotNearHeapLimit()),
+        heapLimitMultiplier(limits.getHeapLimitMultiplier()),
+        heapLimitExceedsMax(limits.getHeapLimitExceedsMax()),
+        heapInitialYoungGenSizeMb(limits.getHeapInitialYoungGenSizeMb()) {}
+
+  v8::Isolate::CreateParams getCreateParams() override {
+    v8::Isolate::CreateParams params;
+    uint64_t softLimit = softHeapLimitMb * 1024 * 1024;
+    if (softLimit > 0) {
+      params.constraints.set_max_young_generation_size_in_bytes(
+          kj::min(softLimit, heapInitialYoungGenSizeMb * 1024 * 1024));
+      params.constraints.set_max_old_generation_size_in_bytes(softLimit);
+      params.array_buffer_allocator_shared =
+          std::make_shared<LimitedArrayBufferAllocator>(softLimit);
+    }
+    return params;
+  }
+
+  static size_t nearHeapLimit(void* data, size_t currentHeapLimit, size_t initialHeapLimit) {
+    auto& self = *static_cast<ConfiguredIsolateLimitEnforcer*>(data);
+    size_t newLimit = currentHeapLimit * self.heapLimitMultiplier;
+    // We can hit this again when taking the heapsnapshot... just increase the limit
+    // and continue, even if it exceeds the configured hard limit
+    if (self.inNearLimitCallback) return newLimit;
+    self.inNearLimitCallback = true;
+    KJ_DEFER(self.inNearLimitCallback = false);
+
+    if ((self.exceededCounter >= self.heapLimitExceedsMax) ||
+        (self.heapHardLimitMb > self.softHeapLimitMb &&
+         newLimit > self.heapHardLimitMb * 1024 * 1024)) {
+      KJ_LOG(ERROR, "Exceeded the configured hard heap limit.",
+             currentHeapLimit,
+             self.exceededCounter,
+             self.heapHardLimitMb);
+      self.maybeGenerateHeapshot();
+      self.v8Isolate->TerminateExecution();
```

Comment:
This will only terminate any currently-executing code, but will not cause the isolate to shut down or be replaced. For workerd, it probably makes more sense to kill the whole process here.

Reply:
Yeah I was considering different ways of handling this and shutting down the whole process after the final snapshot is taken would make sense.